[Merged by Bors] - feat(Geometry/Euclidean/Angle/Unoriented): triangle inequality for angles

LessnessRandomness opened the pull request:
This PR continues the work from #24206.

Original PR: https://github.com/leanprover-community/mathlib4/pull/24206

LessnessRandomness committed:
Triangle inequality for angles

LessnessRandomness committed:
Update TriangleInequality.lean

Some comnents added

LessnessRandomness committed:
Update Mathlib.lean

ran 'lake exe mk_all'

LessnessRandomness committed:
Remove unused import

LessnessRandomness committed:
many theorems to lemmas; some docstrings

LessnessRandomness committed:
Update Basic.lean

Remove "TODO" from docstring and also remove "proof_wanted".

Timeroot committed:
add EuclideanGeometry.angle_triangle

Timeroot committed:
cleanup

Timeroot committed:
rename

Timeroot committed:
attr

LessnessRandomness commented:
## Comments from Original PR #24206

*This section contains 10 comment(s) from the original PR, excluding bot comments.*

---

**@github-actions** ([2025-04-19 19:02 UTC](https://github.com/leanprover-community/mathlib4/pull/24206#issuecomment-2816827960)):
### PR summary [6a858d8eb6](https://github.com/leanprover-community/mathlib4/pull/24206/commits/b5b09dcf8b719cb9eb4b51cbc7dc226a18b2df58)

#### Import changes for modified files

No significant changes to the import graph


<details><summary>Import changes for all files</summary>

|Files|Import difference|
|-|-|
| `Mathlib.Geometry.Euclidean.Angle.Unoriented.TriangleInequality` (new file)|1798|

</details>

---

<details><summary>

#### Declarations diff

</summary>

`+ InnerProductGeometry.angle_triangle`
`+ angle_triangle`
`+ angle_triangle_aux2`
`+ angle_unit_left`
`+ angle_unit_right`
`+ inner_eq_neg_sq_norm_iff`
`+ inner_eq_sq_norm_iff`
`+ inner_orthoDir_nonneg`
`+ inner_orthoDir_zero`
`+ inner_product_of_units_as_cos`
`+ inner_sub_smul_eq_zero`
`+ neg_one_le_inner_unit_unit`
`+ norm_of_unit`
`+ orthoDir`
`+ orthoDir_aux_1`
`+ orthoDir_unit_left`
`+ orthoDir_unit_right`
`+ orthoDir_zero`
`+ sin_as_inner_product`
`+ unit`
`+ unit_eq_id_of_norm_one`
`+ unit_smul_of_pos`
`+ unit_unit`
`+ unit_zero`
`+ zero_orthoDir`
<details>
  <summary>You can run this locally as follows</summary>


```bash
## summary with just the declaration names:
./scripts/declarations_diff.sh <optional_commit>

## more verbose report:
./scripts/declarations_diff.sh long <optional_commit>
```
</details>

The doc-module for `script/declarations_diff.sh` contains some details about this script.

</details>

---

<details><summary>No changes to technical debt.</summary>

You can run this locally as
```
./scripts/technical-debt-metrics.sh pr_summary
```
* The `relative` value is the weighted *sum* of the differences with weight given by the *inverse* of the current value of the statistic.
* The `absolute` value is the `relative` value divided by the total sum of the inverses of the current values (i.e. the weighted *average* of the differences).
</details>

---

**@urkud** ([2025-04-28 05:21 UTC](https://github.com/leanprover-community/mathlib4/pull/24206#issuecomment-2834011563)):
There is a competing PR #22265

---

**@urkud** ([2025-04-28 05:24 UTC](https://github.com/leanprover-community/mathlib4/pull/24206#issuecomment-2834015773)):
Please coordinate with @Timeroot on how to merge your PRs into 1 PR that gets reviewed&merged.

---

**@LessnessRandomness** ([2025-05-03 19:54 UTC](https://github.com/leanprover-community/mathlib4/pull/24206#issuecomment-2848782386)):
Right now **Timeroot** is very busy, we will talk/coordinate when they have free time (week or two later).

---

**@Timeroot** ([2025-05-09 19:06 UTC](https://github.com/leanprover-community/mathlib4/pull/24206#issuecomment-2867646536)):
Alright, I think Ilmārs's is sufficiently clean in terms of the imports that I don't have much to add from my proof - I did the add the `EuclideanGeometry.angle_triangle` version of the theorem, and did some (slightly opinionated) cleanup, hope that was welcome. :)

---

**@LessnessRandomness** ([2025-05-13 12:43 UTC](https://github.com/leanprover-community/mathlib4/pull/24206#issuecomment-2876363059)):
@jsm28 Hello! Busy right now :( will fix the problems starting from the weekend.

Thank you for your work!

---

**@LessnessRandomness** ([2025-05-20 22:29 UTC](https://github.com/leanprover-community/mathlib4/pull/24206#issuecomment-2895975934)):
Ok, so... I talked to @Timeroot and he agreed to take over this PR. I'm just too busy. :(

---

**@Timeroot** ([2025-06-10 02:15 UTC](https://github.com/leanprover-community/mathlib4/pull/24206#issuecomment-2957491210)):
I added my cleanup work, I have some things I'm not sure about regarding whether putting `unit` (or this other function, `orthoDir`) elsewhere in mathlib.

@jsm28 , @LessnessRandomness said he might have some time again, can you advise on the best ways to continue?

---

**@LessnessRandomness** ([2025-06-10 22:53 UTC](https://github.com/leanprover-community/mathlib4/pull/24206#issuecomment-2960761271)):
I will have time in July (probably all the month) or about three weeks from now.

---

**@LessnessRandomness** ([2025-06-18 23:16 UTC](https://github.com/leanprover-community/mathlib4/pull/24206#issuecomment-2986007980)):
Big thank you to @Timeroot!

It has happened that I'm free already.
@jsm28, what should I do next?

jsm28 commented:
Next I suggest moving the more generally meaningful definitions and lemmas about vectors in inner product spaces, or in real normed spaces in general in the case of `unit` and many of its lemmas, to other files (existing or new), probably in a series of separate PRs that this one can then depend on. Note the point from previous discussion that `unit` (meaningful in any real normed space) should probably be renamed, and take care about the naming convention for these lemmas in general.

LessnessRandomness committed:
rename according to Mathlib standards

LessnessRandomness committed:
fixed angle_normalized_left (linter was complaining)

LessnessRandomness commented:
`unit` is renamed to `normalized`.
Also did some renaming, trying to follow the Mathlib standards.

Now thinking about where and how to put different lemmas and defs into different files.

LessnessRandomness committed:
draft (tried to put things into different files)

LessnessRandomness committed:
fixed three linter errors; run `lake exe mk_all`

LessnessRandomness committed:
tried to fix build error (from CI)

LessnessRandomness committed:
Trying to fix linter warnings

LessnessRandomness committed:
Update TriangleInequality.lean

LessnessRandomness committed:
Some improvements

LessnessRandomness committed:
Update Mathlib.lean

Forgot to run `lake exe mk_all`

LessnessRandomness committed:
Renamed UnitSphereAngles to UnitVectorAngles

LessnessRandomness committed:
Docstrings

Added docstring to `Normalized.lean`; edited docstring of `UnitVectorAngles.lean`

LessnessRandomness committed:
Merge branch 'master' into triangle_theorem_for_angles

LessnessRandomness committed:
minor edits

LessnessRandomness committed:
Merge branch 'triangle_theorem_for_angles' of https://github.com/LessnessRandomness/mathlib4 into triangle_theorem_for_angles

LessnessRandomness commented:
@jsm28 I believe I did all I could.

In the file `UnitVectorAngles.lean` - is it enough to have namespace  (?) `UnitVectorAngles`? Or maybe I still should add `_of_norm_one` where it is appropriate?

jsm28 commented on Mathlib/Analysis/NormedSpace/Normalized.lean:
```diff
@@ -0,0 +1,64 @@
+/-
+Copyright (c) 2020 Ilmārs Cīrulis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ilmārs Cīrulis, Alex Meiburg
+-/
+import Mathlib.Analysis.RCLike.Basic
+
+/-!
+# Normalized vector
+
+Function that calculates unit length vector from a vector
+(if the given vector is nonzero vector) or returns zero vector
+(if the given vector is zero vector).
+-/
+
```
I encourage putting this file in a separate PR (which this PR can then depend on) since it's about real normed spaces in general, nothing specifically Euclidean.

LessnessRandomness replied:
Made a separate PR: https://github.com/leanprover-community/mathlib4/pull/26308

LessnessRandomness replied:
How can I make this PR to depend on that separate PR?

jsm28 replied:
The default PR template has a comment explaining how to express dependencies in the PR description.

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,61 @@
+variable [NormedAddCommGroup V]
+variable [InnerProductSpace ℝ V]
+
+@[simp]
+private lemma angle_normalized_left (x y : V) :
+    angle (normalized x) y = angle x y := by
+  by_cases hx : x = 0
+  · simp [hx]
+  replace hx : 0 < ‖x‖⁻¹ := by simp [hx]
+  simp only [normalized, angle_smul_left_of_pos, hx, angle_smul_right_of_pos]
+
+@[simp]
+private lemma angle_normalized_right (x y : V) :
+    angle x (normalized y) = angle x y := by
+  rw [angle_comm, angle_normalized_left, angle_comm]
```
I think these belong in `Mathlib.Geometry.Euclidean.Angle.Unoriented.Basic`, not marked `private`; they're generally meaningful lemmas about angles, not specific to this proof.

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,61 @@
+
+
+/-- **Triangle inequality** for angles between vectors. -/
+theorem InnerProductGeometry.angle_triangle (x y z : V) : angle x z ≤ angle x y + angle y z := by
```
`angle_le_angle_add_angle` would be more like the mathlib naming style (likewise for the other version of this lemma).

LessnessRandomness replied:
Fixed.

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,61 @@
+  [NormedAddTorsor V P]
+
+/-- **Triangle inequality** for angles in Euclidean geometry. -/
+theorem angle_triangle (p p₁ p₂ p₃ : P) : ∠ p₁ p p₃ ≤ ∠ p₁ p p₂ + ∠ p₂ p p₃ :=
```
Also `angle_le_angle_add_angle`.

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/UnitVectorAngles.lean:
```diff
@@ -0,0 +1,171 @@
+variable [InnerProductSpace ℝ V]
+
+
+@[simp]
+lemma inner_self_eq_one {x : V} (hx : ‖x‖ = 1) : ⟪x, x⟫ = 1 :=
+  (inner_eq_one_iff_of_norm_one hx hx).mpr rfl
+
+lemma neg_one_le_inner {x y : V} (hx : ‖x‖ = 1) (hy : ‖y‖ = 1) : -1 ≤ ⟪x, y⟫ := by
+  have H := neg_le_of_abs_le (abs_real_inner_le_norm x y)
+  rw [hx, hy] at H
+  norm_num at H
+  exact H
+
+lemma neg_one_le_inner_normalized_normalized (x y : V) :
+    (-1 : ℝ) ≤ ⟪normalized x, normalized y⟫ := by
```
These all seem like general lemmas that belong somewhere under `Mathlib.Analysis.InnerProductSpace`.

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/UnitVectorAngles.lean:
```diff
@@ -0,0 +1,171 @@
+    simpa [norm_normalized_eq_one_iff]
+  exact neg_one_le_inner H H0
+
+/-- Gets the orthogonal direction of one vector relative to another. -/
+noncomputable def ortho (x y : V) : V := x - ⟪x, y⟫ • y
```
It's probably reasonable to put this and its lemmas (not involving angles) somewhere under `Mathlib.Analysis.InnerProductSpace` as well.

Timeroot replied:
Actually -- it might be nicer to swap the order of the arguments, so that is `ortho y x`. Then `ortho y` is a LinearMap, which is nice.

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/UnitVectorAngles.lean:
```diff
@@ -0,0 +1,171 @@
+  ring
+
+
+lemma inner_eq_cos_angle {x y : V} (hx : ‖x‖ = 1) (hy : ‖y‖ = 1) :
+    ⟪x, y⟫ = Real.cos (angle x y) := by
+  simp [cos_angle, hx, hy]
+
+lemma inner_ortho_right_eq_sin_angle {x y : V} (hx : ‖x‖ = 1) (hy : ‖y‖ = 1):
+    ⟪x, normalized (ortho x y)⟫ = Real.sin (angle x y) := by
```
These probably belong in `Mathlib.Geometry.Euclidean.Angle.Unoriented.Basic`.

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/UnitVectorAngles.lean:
```diff
@@ -0,0 +1,171 @@
+  rw [H0]
+  field_simp
+
+lemma angle_triangle {x y z : V}
+    (Hx : ‖x‖ = 1) (Hy : ‖y‖ = 1) (Hz : ‖z‖ = 1):
```
Also `angle_le_angle_add_angle`, or `angle_le_angle_add_angle_of_norm_one`.

LessnessRandomness replied:
Changed to `angle_le_angle_add_angle`.

LessnessRandomness committed:
Some minor changes

JovanGerb commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/UnitVectorAngles.lean:
```diff
@@ -0,0 +1,170 @@
+    simpa [norm_normalized_eq_one_iff]
+  exact neg_one_le_inner H H0
+
+/-- Gets the orthogonal direction of one vector relative to another. -/
+noncomputable def ortho (y x : V) : V := x - ⟪x, y⟫ • y
```
Can you add to the doc-string that this definition only makes sense if `‖y‖ = 1`.

Should it and the lemmas about it be `private`, as it is only meant for proving lemmas in this file?

LessnessRandomness replied:
I agree with you.

For now I will wait until changes from https://github.com/leanprover-community/mathlib4/pull/26308 appears in Mathlib (if I understand correctly, they still have to be merged or smth like that 🤔). Then I will fix this problems.

LessnessRandomness replied:
As the #26308 is in Matlib now, I started to update the relevant files. Including these two suggested changes. :)

LessnessRandomness committed:
Merge remote-tracking branch 'upstream/master' into triangle_theorem_for_angles

LessnessRandomness committed:
Merge remote-tracking branch 'upstream/master' into triangle_theorem_for_angles

LessnessRandomness committed:
Update relevant files

LessnessRandomness commented:
@JovanGerb Maybe you can tell - what's the current consensus about using `simp only` everywhere?
Should I replace all `simp` with it (with output of `simp?`).

Other than that - everything compiles and looks good to me.

JovanGerb commented:
The rule is generally to use `simp` if it is a terminal tactic, as this is more readable, and `simp only` when it is somewhere in the middle of a tactic sequence, as this is required to be robust. So no need to replace all `simp` with the output of `simp?`.

LessnessRandomness committed:
Remove file UnitVectorAngles (put contents into TriangleInquality)

LessnessRandomness committed:
Change namespace to section

LessnessRandomness commented:
1) I believe that lemmas about unit length vectors have no independent value, so in the end I decided to put them in the same file as the main result (in section named UnitVectorAngles). Am I right to do this?

2) The definition private noncomputable def ortho (y x : V) : V := x - ⟪x, y⟫ • y at Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean gives an useful vector and is meant only for unit length vector y. It can be thought as a part of Gram-Schmidt ortogonalization process, I believe, but I don't need the full Gram-Schmidt, imo, so I hope that this definition is okay.

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,190 @@
+  ring
+
+
+lemma inner_eq_cos_angle {x y : V} (hx : ‖x‖ = 1) (hy : ‖y‖ = 1) :
+    ⟪x, y⟫ = Real.cos (angle x y) := by
+  simp_all [cos_angle]
```
This belongs in some general angle file, it's not specific to the triangle inequality.

LessnessRandomness replied:
Moved it.

LessnessRandomness commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,190 @@
+
+section UnitVectorAngles
+
+@[simp]
+lemma inner_self_eq_one {x : V} (hx : ‖x‖ = 1) : ⟪x, x⟫ = 1 :=
+  (inner_eq_one_iff_of_norm_one hx hx).mpr rfl
+
+lemma neg_one_le_inner {x y : V} (hx : ‖x‖ = 1) (hy : ‖y‖ = 1) : -1 ≤ ⟪x, y⟫ := by
+  have H := neg_le_of_abs_le (abs_real_inner_le_norm x y)
+  simp_all
+
+lemma neg_one_le_inner_normalize_normalize (x y : V) :
+    (-1 : ℝ) ≤ ⟪normalize x, normalize y⟫ := by
+  by_cases hx : x = 0
+  · simp_all
+  by_cases hy : y = 0
+  · simp_all
+  have H: ‖normalize x‖ = 1 := norm_normalize_eq_one_iff.mpr hx
+  have H0: ‖normalize y‖ = 1 := norm_normalize_eq_one_iff.mpr hy
+  exact neg_one_le_inner H H0
```
Thank you, will think where exactly to put them.

LessnessRandomness replied:
Moved the first two, inlined the third one.

LessnessRandomness committed:
Merge remote-tracking branch 'upstream/master' into triangle_theorem_for_angles

LessnessRandomness committed:
Move few lemmas to better places

LessnessRandomness committed:
Remove unnecessary argument of simp (Geometry\Manifold\Instances\Sphere)

LessnessRandomness committed:
Merge branch 'master' into triangle_theorem_for_angles

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+private noncomputable def ortho (y x : V) : V := x - ⟪x, y⟫ • y
+
+private lemma inner_ortho_nonneg {x y : V} (hx : ‖x‖ = 1) (hy : ‖y‖ = 1) : 0 ≤ ⟪x, ortho y x⟫ := by
+  rw [ortho, inner_sub_right, real_inner_smul_right, inner_self_eq_one_of_norm_one hx]
+  rw [← sq, sub_nonneg, sq_le_one_iff_abs_le_one]
+  have H := abs_real_inner_le_norm x y
+  simp_all
```
```suggestion
  rw [ortho, inner_sub_right, inner_self_eq_one_of_norm_one hx, real_inner_smul_right, ← sq]
  grw [sub_nonneg, sq_le_one_iff_abs_le_one, abs_real_inner_le_norm, hx, hy, one_mul]
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+
+private lemma inner_normalize_ortho (x : V) {y : V} (hy : ‖y‖ = 1) :
+    ⟪y, normalize (ortho y x)⟫ = 0 := by
+  rw [ortho, NormedSpace.normalize, real_inner_smul_right, inner_sub_right,
+      real_inner_smul_right, inner_self_eq_one_of_norm_one hy, real_inner_comm]
+  simp
```
```suggestion
  simp [ortho, NormedSpace.normalize, real_inner_smul_right, inner_sub_right,
    inner_self_eq_one_of_norm_one hy, real_inner_comm]
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+  · simp_all
+  rw [real_inner_self_eq_norm_sq, hx]
+  have H1 : ‖x - ⟪x, y⟫ • y‖ ≠ 0 := by
+    simp only [ne_eq, norm_eq_zero];
+    intro H2; rw [sub_eq_zero] at H2
+    rw [H2, norm_smul, hy] at hx
+    simp only [Real.norm_eq_abs, mul_one] at hx
```
```suggestion
    simp only [ne_eq, norm_eq_zero, sub_eq_zero]
    intro H2
    rw [H2, norm_smul, hy, Real.norm_eq_abs, mul_one] at hx
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+    apply eq_or_eq_neg_of_abs_eq at hx
+    rcases hx with hx | hx <;> rw [hx] at H2 <;> simp_all
+  field_simp
+  rw [← real_inner_self_eq_norm_sq]
+  rw [inner_sub_left, inner_sub_right, inner_sub_right]
+  rw [real_inner_smul_right, real_inner_smul_left]
+  rw [real_inner_smul_right, real_inner_smul_left]
+  rw [real_inner_comm x y]
+  rw [real_inner_self_eq_norm_sq, hx, real_inner_self_eq_norm_sq, hy]
```
you shouldn't separate rewrites like this
```suggestion
  rw [← real_inner_self_eq_norm_sq, inner_sub_left, inner_sub_right, inner_sub_right,
    real_inner_smul_right, real_inner_smul_left, real_inner_smul_right, real_inner_smul_left,
    real_inner_comm x y, real_inner_self_eq_norm_sq, hx, real_inner_self_eq_norm_sq, hy]
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+  rw [sq_eq_sq_iff_abs_eq_abs, abs_of_nonneg (sin_angle_nonneg x y)] at H
+  have H0 : 0 ≤ ⟪x, normalize (ortho y x)⟫ := by
+    rw [NormedSpace.normalize, real_inner_smul_right]
+    have H1 := inner_ortho_nonneg hx hy
+    have H2 := norm_nonneg (ortho y x)
+    have H3 := inv_nonneg_of_nonneg H2
+    exact Left.mul_nonneg H3 H1
```
```suggestion
    exact Left.mul_nonneg (inv_nonneg_of_nonneg (norm_nonneg (ortho y x)))
      (inner_ortho_nonneg hx hy)
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+  · simpa [hy] using angle_le_pi x z
+  by_cases hz : z = 0
+  · simpa [hz] using angle_nonneg x y
+  have H := angle_le_angle_add_angle_of_norm_one
+    (norm_normalize_eq_one_iff.mpr hx) (norm_normalize_eq_one_iff.mpr hy)
+    (norm_normalize_eq_one_iff.mpr hz)
+  simp at H
+  exact H
```
```suggestion
  simpa using angle_le_angle_add_angle_of_norm_one (norm_normalize_eq_one_iff.mpr hx)
    (norm_normalize_eq_one_iff.mpr hy) (norm_normalize_eq_one_iff.mpr hz)
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+  simp at H
+  exact H
+
+
```
```suggestion
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+  have H0 : 0 ≤ angle x y + angle y z :=
+    add_nonneg (angle_nonneg x y) (angle_nonneg y z)
+  have H1 : ⟪x, z⟫ = ⟪x, z⟫ := rfl
+  nth_rw 2 [angle_le_angle_add_angle_aux Hx Hy] at H1
+  nth_rw 2 [angle_le_angle_add_angle_aux Hz Hy] at H1
```
```suggestion
  nth_rw 2 [angle_le_angle_add_angle_aux Hx Hy, angle_le_angle_add_angle_aux Hz Hy] at H1
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+  have H1 : ⟪x, z⟫ = ⟪x, z⟫ := rfl
+  nth_rw 2 [angle_le_angle_add_angle_aux Hx Hy] at H1
+  nth_rw 2 [angle_le_angle_add_angle_aux Hz Hy] at H1
+  simp only [inner_add_left, inner_add_right, real_inner_smul_left, real_inner_smul_right] at H1
+  simp only [real_inner_comm y (normalize _), real_inner_self_eq_norm_sq, Hy,
+    angle_comm z y, inner_normalize_ortho] at H1
```
```suggestion
  simp only [inner_add_left, inner_add_right, real_inner_smul_left, real_inner_smul_right,
    real_inner_comm y (normalize _), real_inner_self_eq_norm_sq, Hy,
    angle_comm z y, inner_normalize_ortho] at H1
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+      (norm_normalize_eq_one_iff.mpr H3) (norm_normalize_eq_one_iff.mpr H4)
+  have H3 := mul_nonneg (sin_angle_nonneg x y) (sin_angle_nonneg y z)
+  have H4 : Real.cos (angle x y + angle y z) ≤ Real.cos (angle x z) := by
+    rw [Real.cos_add, ← inner_eq_cos_angle_of_norm_one Hx Hz]
+    rw [neg_le_iff_add_nonneg] at H2
+    rw [H1]
+    field_simp
+    linarith [mul_nonneg H3 H2]
```
```suggestion
    rw [neg_le_iff_add_nonneg] at H2
    rw [Real.cos_add, ← inner_eq_cos_angle_of_norm_one Hx Hz, H1]
    linarith [mul_nonneg H3 H2]
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+  rw [← inner_ortho_right_eq_sin_angle Hx Hy, ← inner_eq_cos_angle_of_norm_one Hx Hy]
+  by_cases hxy : x - ⟪x, y⟫ • y = 0
+  · simp [ortho, hxy, ← sub_eq_zero]
+  simp only [ortho]
+  rw [NormedSpace.normalize, real_inner_smul_right, inner_sub_right, real_inner_smul_right]
+  rw [real_inner_self_eq_norm_sq, Hx, ← sq, mul_smul, ← smul_assoc]
```
```suggestion
  rw [ortho, NormedSpace.normalize, real_inner_smul_right, inner_sub_right, real_inner_smul_right,
    real_inner_self_eq_norm_sq, Hx, ← sq, mul_smul, ← smul_assoc]
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+    rw [sub_ne_zero, ne_comm, sq_ne_one_iff]
+    constructor <;> contrapose! hxy
+    · rw [inner_eq_one_iff_of_norm_one Hx Hy] at hxy
+      simp_all [inner_self_eq_one_of_norm_one]
```
```suggestion
      simp [Hy, hxy, inner_self_eq_one_of_norm_one]
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+    · rw [inner_eq_one_iff_of_norm_one Hx Hy] at hxy
+      simp_all [inner_self_eq_one_of_norm_one]
+    · rw [inner_eq_neg_one_iff_of_norm_one Hx Hy] at hxy
+      simp_all [inner_self_eq_one_of_norm_one]
```
```suggestion
      simp [Hy, hxy, inner_self_eq_one_of_norm_one]
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+  field_simp
+  rw [sq, ← real_inner_self_eq_norm_sq]
+  have H0 : ⟪x - ⟪x, y⟫ • y, x - ⟪x, y⟫ • y⟫ = 1 - ⟪x, y⟫ ^ 2 := by
+    rw [inner_sub_left, inner_sub_right, inner_sub_right, inner_self_eq_one_of_norm_one Hx]
+    rw [real_inner_smul_right, ← sq]
+    field_simp
+    rw [real_inner_smul_left, real_inner_smul_left, real_inner_smul_right]
+    rw [inner_self_eq_one_of_norm_one Hy, real_inner_comm y x]
```
```suggestion
    rw [inner_sub_left, inner_sub_right, inner_sub_right, inner_self_eq_one_of_norm_one Hx,
      real_inner_smul_right, ← sq, real_inner_smul_left, real_inner_smul_left,
      real_inner_smul_right, inner_self_eq_one_of_norm_one Hy, real_inner_comm y x]
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+variable [NormedAddCommGroup V]
+variable [InnerProductSpace ℝ V]
+
+
```
```suggestion
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,172 @@
+
+end UnitVectorAngles
+
+
+
```
```suggestion
```

LessnessRandomness committed:
apply changes suggested by themathqueen

LessnessRandomness committed:
try to fix linter error in Analysis/InnerProductSpace/PiL2

LessnessRandomness commented on Mathlib/Analysis/InnerProductSpace/PiL2.lean:
```diff
@@ -422,7 +422,7 @@ lemma inner_eq_zero (b : OrthonormalBasis ι 𝕜 E) {i j : ι} (hij : i ≠ j)
 
 @[simp]
```
Thank you!

themathqueen replied:
there's also another one in `Mathlib/Analysis/Fourier/FiniteAbelian/Orthogonality.lean:45:48`

LessnessRandomness committed:
Fix linter error

LessnessRandomness committed:
Fix linter error in Analysis/Fourier/FiniteAbelian/Orthogonality

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/Basic.lean:
```diff
@@ -336,4 +329,25 @@ theorem sin_eq_one_iff_angle_eq_pi_div_two : sin (angle x y) = 1 ↔ angle x y =
+@[simp]
+lemma angle_normalize_right (x y : V) :
+    angle x (NormedSpace.normalize y) = angle x y := by
+  rw [angle_comm, angle_normalize_left, angle_comm]
```
```suggestion
  simp [angle_comm x]
```

themathqueen commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,151 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ilmārs Cīrulis, Alex Meiburg
+-/
+import Mathlib.Geometry.Euclidean.Angle.Unoriented.Basic
+import Mathlib.Geometry.Euclidean.Angle.Unoriented.Affine
+import Mathlib.Analysis.NormedSpace.Normalize
```
Reorganize the imports alphabetically please

LessnessRandomness committed:
define ortho using Submodule.starProjection

LessnessRandomness committed:
apply suggestions by themathqueen

LessnessRandomness committed:
change order of import to alphabetic in Geometry/Euclidean/Unoriented/Bassic too

LessnessRandomness committed:
Merge remote-tracking branch 'upstream/master' into triangle_theorem_for_angles

LessnessRandomness committed:
Fix spelling error

JovanGerb commented on Mathlib/Analysis/InnerProductSpace/Basic.lean:
```diff
@@ -15,7 +15,7 @@ This file proves many basic properties of inner product spaces (real or complex)
 
 - `inner_mul_inner_self_le`: the Cauchy-Schwarz inequality (one of many variants).
-- `norm_inner_eq_norm_iff`: the equality criterion in the Cauchy-Schwarz inequality (also in many
+- `norm_inner_eq_norm_iff`: the equality criteion in the Cauchy-Schwartz inequality (also in many
```
What happened here? Why is it now "criteion"?

LessnessRandomness replied:
Great, I missed the second "Schwartz" that had to be changed to "Schwarz", too.

LessnessRandomness committed:
Fix spelling error once again (Schwartz -> Schwarz)

LessnessRandomness committed:
Merge branch 'master' into triangle_theorem_for_angles

LessnessRandomness commented on Mathlib/Analysis/InnerProductSpace/Basic.lean:
```diff
@@ -796,6 +796,25 @@ theorem inner_eq_one_iff_of_norm_one {x y : E} (hx : ‖x‖ = 1) (hy : ‖y‖
+  simpa [hx, hy] using neg_le_of_abs_le (abs_real_inner_le_norm x y)
+
+@[simp]
+theorem inner_self_eq_one_of_norm_one {x : E} (hx : ‖x‖ = 1) : ⟪x, x⟫_𝕜 = 1 :=
```
#30793 - here I try to benchmark a PR with this simp lemma.

Though it failed: https://github.com/leanprover-community/mathlib4/actions/runs/18729718862/job/53423491132?pr=30793
With two errors that I don't know how to fix.

LessnessRandomness committed:
Merge branch 'master' into triangle_theorem_for_angles

LessnessRandomness committed:
Remove a simp tag

LessnessRandomness committed:
Remove unused argument in two simp calls

LessnessRandomness committed:
Two changes

LessnessRandomness committed:
Add a theorem

LessnessRandomness commented on Mathlib/Analysis/InnerProductSpace/Basic.lean:
```diff
@@ -793,6 +793,21 @@ theorem inner_eq_one_iff_of_norm_one {x y : E} (hx : ‖x‖ = 1) (hy : ‖y‖
+theorem neg_one_le_real_inner_of_norm_one {x y : F} (hx : ‖x‖ = 1) (hy : ‖y‖ = 1) :
+    -1 ≤ ⟪x, y⟫_ℝ := by
+  simpa [hx, hy] using neg_le_of_abs_le (abs_real_inner_le_norm x y)
+
```
Decided to add it. If anything, it can be always removed.

JovanGerb commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/Affine.lean:
```diff
@@ -18,9 +18,6 @@ This file defines unoriented angles in Euclidean affine spaces.
 * `EuclideanGeometry.angle`, with notation `∠`, is the undirected angle determined by three
   points.
 
```
I think another newline should be removed here?

JovanGerb commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/Basic.lean:
```diff
@@ -336,4 +329,25 @@ theorem sin_eq_one_iff_angle_eq_pi_div_two : sin (angle x y) = 1 ↔ angle x y =
+@[simp]
+lemma angle_normalize_right (x y : V) :
+    angle x (NormedSpace.normalize y) = angle x y := by
+  simp [angle_comm x]
```
If you do `rw [angle_comm]; simp`, you can avoid needing to pass an argument to the lemma.

LessnessRandomness replied:
Replaced the line with `rw [angle_comm, angle_normalize_left, angle_comm]`.
(`rw [angle_comm]; simp` didn't close the goal.)

JovanGerb commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,156 @@
+    rw [H2, norm_smul, hy, Real.norm_eq_abs, mul_one] at hx
+    apply eq_or_eq_neg_of_abs_eq at hx
+    rcases hx with hx | hx <;> rw [hx] at H2 <;> simp_all
+  field_simp
```
I forgot if we have a policy against using `field_simp` followed by `rw` (because `field_simp` has a flexible output, and `rw` is inflexible about when it works). Can this be refactored to avoid this?

JovanGerb commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,156 @@
+
+variable {V : Type*}
+variable [NormedAddCommGroup V]
+variable [InnerProductSpace ℝ V]
```
These can go on just one line

JovanGerb reviewed (commented):
I've left a few small comments.

themathqueen commented:
> Maybe the theorem names should end in `of_norm_eq_one` instead of `of_norm_one`?

I agree with this.

(Note that there are currently more instances of `_of_norm_one` than `_of_norm_eq_one` in `Mathlib`, but I'll PR something tonight.)

LessnessRandomness committed:
Change of_norm_one to of_norm_eq_one in theorem names

LessnessRandomness committed:
Some changes

themathqueen commented on Mathlib/Analysis/Complex/Norm.lean:
```diff
@@ -354,7 +354,7 @@ lemma re_neg_ne_zero_of_re_pos {s : ℂ} (hs : 0 < s.re) : (-s).re ≠ 0 :=
   re_neg_ne_zero_of_re_pos <| zero_lt_one.trans hs
 
-lemma norm_sub_one_sq_eq_of_norm_one {z : ℂ} (hz : ‖z‖ = 1) :
+lemma norm_sub_one_sq_eq_of_norm_eq_one {z : ℂ} (hz : ‖z‖ = 1) :
```
I've already done these: #31674.
Just rename the ones you are adding. Otherwise, it's too many changes.

LessnessRandomness replied:
Okay

LessnessRandomness committed:
Fix linter error and also an unsolved goal

LessnessRandomness committed:
Revert "Fix linter error and also an unsolved goal"

This reverts commit d4b35256d331b2a909388885c1cfb75bd32f136d.

LessnessRandomness committed:
Revert "Some changes"

This reverts commit d2a5f70d30a34c093b6cb888dcfbd93a85bb4eca.

LessnessRandomness committed:
Revert "Change of_norm_one to of_norm_eq_one in theorem names"

This reverts commit 5f793e5861183193089f1637315fa480ee31e94b.

LessnessRandomness committed:
Some fixes

LessnessRandomness committed:
Change "of_norm_one" to "of_norm_eq_one" (one file)

LessnessRandomness committed:
Tiny refactor

LessnessRandomness committed:
Merge branch 'master' into triangle_theorem_for_angles

LessnessRandomness committed:
Fix

dagurtomas committed:
Merge branch 'master' into triangle_theorem_for_angles

dagurtomas committed:
Apply suggestion from @dagurtomas

JovanGerb commented:
Since most declarations in the new file are currenctly marked `private`, it would make more sense to not use `@[expose] public section`, but instead to remove all occurrences of `private`, and to mark the two versions of `angle_le_angle_add_angle` as `public`.

LessnessRandomness committed:
Apply suggestion from @JovanGerb

JovanGerb commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,158 @@
+  by_cases h₁ : x = y
+  · simp_all
+  by_cases h₂ : x = - y
+  · simp_all
```
```suggestion
  · simp [*]
  by_cases h₂ : x = - y
  · simp [*]
```
`simp [*]` is a less powerful form

JovanGerb commented on Mathlib/Geometry/Euclidean/Angle/Unoriented/TriangleInequality.lean:
```diff
@@ -0,0 +1,158 @@
+    intro H2
+    rw [H2, norm_smul, hy, Real.norm_eq_abs, mul_one] at hx
+    apply eq_or_eq_neg_of_abs_eq at hx
+    rcases hx with hx | hx <;> rw [hx] at H2 <;> simp_all
```
```suggestion
    rcases hx with hx | hx <;> simp [*] at H2

```

JovanGerb reviewed (approved):
Thanks! Just some small comments about the proofs.

LessnessRandomness committed:
Apply changes suggested by @JovanGerb

jcommelin reviewed (commented):
Thanks :tada:

bors merge
